lib/matplotlib/backends/backend_pdf.py

```diff
@@ -2176,52 +2176,46 @@ def draw_text_woven(chunks):
-                            name = self.file._get_xobject_symbol_name(
-                                font.fname, glyph_name)
-                            self.file.output(Name(name), Op.use_xobject)
-                            self.file.output(Op.grestore)
-
-                        # Move the pointer based on the character width
-                        # and kerning
-                        glyph = font.load_char(ccode, flags=LOAD_NO_HINTING)
-                        if lastgind is not None:
-                            kern = font.get_kerning(
-                                lastgind, gind, KERNING_UNFITTED)
-                        else:
-                            kern = 0
-                        lastgind = gind
-                        newx += kern / 64 + glyph.linearHoriAdvance / 65536
-
-                if mode == 1:
-                    self.file.output(Op.end_text)
+            # Output all the 1-byte characters in a BT/ET group.
+            self.file.output(Op.begin_text,
+                             self.file.fontName(prop),
+                             fontsize,
+                             Op.selectfont)
+            newx = oldx = 0
+            for chunk_type, chunk in chunks:
+                if chunk_type == 1:
+                    self._setup_textpos(newx, 0, 0, oldx, 0, 0)
+                    self.file.output(self.encode_string(chunk, fonttype),
+                                     Op.show)
+                    oldx = newx
+                # Update newx to include the advance from this chunk,
+                # regardless of its mode...
+                for char_idx, newx in _text_layout.layout(
+                        chunk, font, x0=newx, kern_mode=KERNING_UNFITTED):
+                    pass
+                newx += (  # ... including the advance from the last character.
+                    font.load_glyph(char_idx, flags=LOAD_NO_HINTING)
+                    .linearHoriAdvance / 65536)
+            self.file.output(Op.end_text)
+
+            # Then output all the 2-byte characters.
+            newx = 0
+            for chunk_type, chunk in chunks:
+                for char_idx, newx in _text_layout.layout(
+                        chunk, font, x0=newx, kern_mode=KERNING_UNFITTED):
+                    if chunk_type == 2:
+                        glyph_name = font.get_glyph_name(char_idx)
+                        self.file.output(Op.gsave)
+                        self.file.output(0.001 * fontsize, 0,
+                                         0, 0.001 * fontsize,
+                                         newx, 0, Op.concat_matrix)
+                        name = self.file._get_xobject_symbol_name(
+                            font.fname, glyph_name)
+                        self.file.output(Name(name), Op.use_xobject)
+                        self.file.output(Op.grestore)
+                newx += (  # ... including the advance from the last character.
```

Comment:
Maybe rename the outer or inner `newx` would be clearer?

Reply:
done, I agree it's a bit better